### PR TITLE
Templated sum when

### PR DESCRIPTION
**Pull request by orangejenny**
https://dimagi-dev.atlassian.net/browse/ICDS-871

See https://github.com/dimagi/commcare-hq/pull/25880 for detail.

### CODE REVIEW

File: sqlagg/columns.py [hunk 1]
```diff
@@ -88,15 +85,37 @@ def sql_column(self):
...
     """
     aggregate_fn = func.sum
 
 
+class SumWhenWithBinds(SumWhen):
+    """
+    SumWhen("age_cohort",
+            whens=[
+                ["when": "age < :ten", "then": 1],
+                ["when": "age < :thirty", "then": 2],
```
Comment (emord) [inline sqlagg/columns.py]: are these supposed to be dictionaries?

File: sqlagg/columns.py [hunk 2]
```diff
@@ -88,15 +85,37 @@ def sql_column(self):
...
+    def sql_column(self):
+        return ConditionalTemplateColumn(self.key, self.whens, self.binds, self.else_,
+                                         self.aggregate_fn, self.alias)
+
+
 class ConditionalColumn(SqlColumn):
     """
     ConditionalColumn("vehicle",
-                      whens={"unicycle": 1, "bicycle": 2, "car": 4},
+                      whens=[["unicycle", 1], ["bicycle": 2], ["car": 4]],
```
Comment (emord) [inline sqlagg/columns.py]: I'm not sure if this should be a list of lists (like the example in docstring of `SumWhen`) or a list of dictionaries

File: sqlagg/columns.py [hunk 3]
```diff
@@ -115,11 +115,18 @@ def build_column(self):
...
             whens = []
             for item in self.whens:
-                when, then = item
-                if isinstance(then, six.string_types):
-                    whens.append((text(when), text(then)))
-                else:
-                    whens.append((text(when), then))
+                when = item[0]
+                params = item[1:-1]
+                then = item[-1]
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply sqlagg/columns.py]: thanks, I like that

**Comment by orangejenny** [comment]
@esoergel @emord Reworked this so that `whens` is always a list, but of variable length: `[expression, *binds, value]`. This removes a bunch of code; It's probably easier to review from the top. The only thing that I don't like is that there isn't a good way to validate the difference between `[expression, p1, value]` and `[expression, value, p1]`.

**Pull request merged**